src/sql/services/notebook/localContentManager.ts

```diff
@@ -183,8 +185,182 @@ namespace v4 {
+			nbformat: 4,
+			nbformat_minor: nbversion.MINOR_VERSION
+		};
+
+		if (contents.worksheets) {
+			for (let worksheet of contents.worksheets) {
+				if (worksheet.cells) {
+					notebook.cells.push(...worksheet.cells.map(cell => createCell(cell)));
+				}
+			}
+		}
+
+		return notebook;
+	}
+
+	function createCell(cell: Cell): nb.ICellContents {
+		switch (cell.cell_type) {
+			case 'markdown':
+			case 'raw':
+				return v4.createDefaultCell(cell);
+			case 'code':
+				return createCodeCell(cell as CodeCell);
+			case 'heading':
+				return createHeadingCell(cell);
+			default:
+				throw new TypeError(`Cell type ${(cell as any).cell_type} unknown`);
+		}
+	}
+
+
+	function createMimeBundle(oldMimeBundle: MimeOutput): MimeBundle {
+		let mimeBundle: MimeBundle = {};
+		for (let key of Object.keys(oldMimeBundle)) {
+			// v3 had non-media types for rich media
+			if (key in VALID_MIMETYPES) {
+				let newKey = VALID_MIMETYPES[key as MimeTypeKey];
+				mimeBundle[newKey] = v4.cleanMimeData(newKey, oldMimeBundle[key]);
+			}
+		}
+		return mimeBundle;
+	}
+
+	const createOutput = (output: Output): nb.ICellOutput => {
+		switch (output.output_type) {
+			case 'pyout':
+				return <nb.IExecuteResult> {
+					output_type: OutputTypes.ExecuteResult,
+					execution_count: output.prompt_number,
+					data: createMimeBundle(output),
+					metadata: output.metadata
+				};
+			case 'display_data':
+				return <nb.IDisplayData> {
+					output_type: OutputTypes.DisplayData,
+					data: createMimeBundle(output),
+					metadata: output.metadata
+				};
+			case 'stream':
+				// Default to stdout in all cases unless it's stderr
+				const name = output.stream === 'stderr' ? 'stderr' : 'stdout';
+				return <nb.IStreamResult> {
+					output_type: OutputTypes.Stream,
+					name: name,
+					text: v4.demultiline(output.text)
+				};
+			case 'pyerr':
+				return <nb.IErrorResult> {
+					output_type: OutputTypes.Error,
+					ename: output.ename,
+					evalue: output.evalue,
+					traceback: output.traceback
+				};
+			default:
+				throw new TypeError(localize('unrecogniedOutputType', 'Output type {0} not recognized', output.output_type));
+		}
+	};
+
+	function createCodeCell(cell: CodeCell): nb.ICellContents {
+		return <nb.ICellContents> {
+			cell_type: cell.cell_type,
+			source: v4.demultiline(cell.input),
+			outputs: cell.outputs.map(createOutput),
+			execution_count: cell.prompt_number,
+			metadata: cell.metadata
+		};
+	}
+
+	function createHeadingCell(cell: HeadingCell): nb.ICellContents {
+		// v3 heading cells are just markdown cells in v4+
+		return <nb.ICellContents> {
+			cell_type: 'markdown',
+			source: Array.isArray(cell.source)
+				? v4.demultiline(
+					cell.source.map(line =>
+						Array(cell.level)
+							.join('#')
+							.concat(' ')
+							.concat(line)
+					)
+				)
+				: cell.source,
+			metadata: cell.metadata
+		};
+	}
+
+	const VALID_MIMETYPES = {
+		text: 'text/plain',
```

Comment:
Not here. This is the entire set of supported types for v3. These were just included in the output object instead of having their own `data` object to be stored in.